### ISSUE TITLE
Exception when attempts count is reached in Prompt

### ISSUE THREAD

**Issue by evnik** [Microsoft/BotBuilder#466]
Why the `Prompt<T, U>` [throws](https://github.com/Microsoft/BotBuilder/blob/9ebdfac5aa7c02d324361c3593cf82c1ca705a60/CSharp/Library/Dialogs/PromptDialog.cs#L599) an `Exception` if attempts count is reached? How caller code should recognize that case? Even the PizzaBot sample fails if you try to enter **Today** instead of day of week three times.
Maybe it'll be better to replace it on call of `Done` passing default value of `T`?
```C#
context.Done(default(T))
```

**Comment by willportnoy** [comment]
In the callback specified when the Prompt is called, the await on the IAwaitable should propagate the exception, so you should be able to detect that case.

If you're suggesting "exceptions as control flow" is not appropriate, that's a reasonable position to take - I don't understand what you mean yet.

**Comment by evnik** [comment]
Well, let me try to explain again:

1. It isn't clear that caller code should surround the `await result;` statement with `try ... catch` to handle the **too many attempts** case in the `resume` handler. (I just found the clue)
1. Thrown exception is of too unspecified type to be clear when it was a logic error or the **too many attempts** case. (You can compare the exception message though, but that's not a good practice)
1. If you forget to use the `try ... catch`, then whole the pipeline fails, so the `await context.PostAsync(this.MakePrompt(context, promptOptions.TooManyAttempts));` actually doesn't make sense. Otherwise, you can post `ex.Message` right from your `catch` block.

So, yes, the bottom line is to avoid using of exception as control flow here. Another option is to use a dedicated exception type and to make it more clear in the comments for `Prompt`.

**Comment by willportnoy** [comment]
I'll go with the "use a dedicated exception type and to make it more clear in the comments for Prompt" options for now - I agree with the "exception as control flow" comment, but that will take a larger change.

**Comment by evnik** [comment]
Well, better than nothing.
Could you clarify please why call of `Done` passing default value of `T` doesn't work? For me that pretty small change should fix the problem without major harm to compartibility.
Do you expect introduction of a default option in the `PromptOptions` for the full fix? If you are able to explain this I'm glad to contribute with implementation PR.

**Comment by willportnoy** [comment]
Done(default(T)) seems like "swallowing the error".  For example, if T were an Int32, and the user typed "a" too many times, should we silently return default(T) which is 0?

For references types, it may be better, though it pushes a TooManyAttemptsException now into a NullReferenceException later.

It seems like the best solution is to have it return an Option type, where there is a clear signal between Success and Failure.

**Comment by evnik** [comment]
That explanation is much more clear and sounds reasonable, thank you. Actually, I didn't take in account any cases of `T` type except bool, string or an enum, and all of them should work with `default(T)`.
So, what do you think, is the team able to push an effort for this in the nearest future? If no, are you guys open for contributions here?

**Comment by willportnoy** [comment]
Which solution are you thinking?  The one based on the Option type?

**Comment by evnik** [comment]
Yes, I assume it is acceptable and looks good to me.

**Comment by willportnoy** [comment]
Sorry for the delays - I've been thinking of the backward compatibility story.  We could have overloads of the Prompt methods, one which takes the delegate with the IAwaitable<T> argument and another that takes the IAwaitable<Option<T>> argument (Option for "expected" failure due to retry exhaustion, IAwaitable for "unexpected" failures) so that a prompt caller could declare their intention to handle retry exhaustion without exceptions.

**Comment by v-thgratMSGit** [comment]
Closing this issue. Thanks!
